torch/utils/_benchmark/utils/timer.py

```diff
@@ -105,26 +109,43 @@ def stop_hook(times):
             return False
         times = self._threaded_measurement_loop(number, time_hook, stop_hook, min_run_time, max_run_time, callback=callback)
         measure = self._construct_measurement(number, times)
+        measure.set_cache_speedup(cache_speedup)
         return measure
 
-    def _estimate_block_size(self, min_run_time):
+    def _measure_uncached_runtime(self):
+        cache_clear = common.CPUCacheClear(cache_size_mb=_CACHE_SENSITIVITY_CACHE_SIZE_MB)
+        cache_clear.clear_cpu_cache()
+        populate_timer = timeit.Timer('sum(range(2)); torch.tanh(torch.tensor([1.1]))',
+                                      globals={'torch': torch})
+        populate_timer.timeit(1)
+        return self._timer.timeit(1)
+
+    def _estimate_block_size(self, min_run_time: float, check_cache_speedup=True):
+        cache_sensitivity = None
+        overhead = np.median([self._timer.timeit(0) for _ in range(5)])
+        cache_speedup = 0.0
+        uncached_time = 0.0
+        if check_cache_speedup:
+            uncached_time = self._measure_uncached_runtime()
+
+        number = 1
         with common.set_torch_threads(self._num_threads):
             # Estimate the block size needed for measurement to be negligible
             # compared to the inner loop. This also serves as a warmup.
-            overhead = np.median([self._timer.timeit(0) for _ in range(5)])
-            number = 1
             while True:
                 time_taken = self._timer.timeit(number)
+                if check_cache_speedup:
+                    cache_speedup = uncached_time / (time_taken / number)
```

Comment:
This approach is not robust, as it implicitly assumes that measured time is linear with iterations. A better approach would be to run the early iterations both with and without a cache clear and compare the two. (In effect checking for an extra constant overhead.)

Reply:
I don't think clearing the cache once then running 10 times is at all meaningful, the clear cache effects are reduced by an order of magnitude. Also, we only warn for a speedup of 1.5; so we would only detect for the N=10 case if the uncached slow down was 5x. Plus, the reported speedup would be misleading and possibly vary (depending on which check we hit it at).

We can compare 1-to-1 makes sense, but I suspect the clear isn't worth the added overhead for the 2nd round with N=10.

Reply:
Happy to run some experiments here if you'd like to dig into it.